Lib/importlib/_bootstrap_external.py

```diff
@@ -91,22 +91,42 @@ def _unpack_uint16(data):
     return int.from_bytes(data, 'little')
 
 
-def _path_join(*path_parts):
-    """Replacement for os.path.join()."""
-    return path_sep.join([part.rstrip(path_separators)
-                          for part in path_parts if part])
+if _MS_WINDOWS:
+    def _path_splitroot(p):
+        return _os._path_splitroot(p.replace('/', path_sep))
+
+    def _path_join(*path_parts):
+        """Replacement for os.path.join()."""
+        if not path_parts:
+            return ""
+        root, tail = _path_splitroot(path_parts[0])
+        path = [tail]
+        for new_root, tail in map(_path_splitroot, path_parts[1:]):
+            if new_root == path_sep:
+                path = [tail]
+            elif new_root.endswith(path_sep):
+                root = new_root[:-1]
+                path = [tail]
+            elif new_root:
+                root = new_root
+                path.append(tail)
+            else:
+                path.append(tail)
+        return path_sep.join(p.rstrip(path_separators) for p in [root, *path] if p)
```

Comment:
`_join_path()` can be made more rigorous with a few minor changes. Example cases:

* `_join_path("//spam/eggs/foo", "//spam/bar")` is "\\\\spam\\bar", not "\\\\spam\\bar\\foo".
* `_join_path("C:", "/")` is "C:\\", not "C:".
* `_join_path("C:spam", "D:eggs")`  is "D:eggs", not "D:\\spam\\eggs".

I know this function has fairly limited applications in this module, which is typically just to append a relative filename to an absolute path. But it may as well be nearly as reliable in Windows as it is in POSIX, since `_path_splitroot()` does some of the more difficult work, in close approximation to `ntpath.splitdrive()`.

Modified version, labeled with the path type(s) in each case:

```python
    def _path_join(*path_parts):
        """Replacement for os.path.join()."""
        if not path_parts:
            return ""
        root, *path = _path_splitroot(path_parts[0])
        for new_root, tail in map(_path_splitroot, path_parts[1:]):
            if new_root == path_sep:
                # Rooted
                if not root.endswith(path_sep):
                    root += path_sep
                path = [tail]
            elif new_root.startswith(path_sep) or new_root.endswith(path_sep):
                # UNC Absolute, Drive Absolute, Local Device
                root = new_root
                path = [tail]
            elif new_root:
                # Drive Relative
                if new_root.upper() == root.rstrip(path_sep).upper():
                    path.append(tail)
                else:
                    root = new_root
                    path = [tail]
            else:
                # Relative
                path.append(tail)
        path = path_sep.join(p.rstrip(path_sep) for p in path if p)
        if path and root.startswith(path_sep) and not root.endswith(path_sep):
            root += path_sep
        return root + path
```
Part of the above change is for `root` to always include the root backslash if it's explicitly present. It may need to be added at the end for the UNC case if the joined `path` is non-empty, but it shouldn't be added for a drive-relative root, such as "D:". 

A logical drive such as "D:" has its own working directory, e.g. `nt._getfullpathname('D:')`, which is not necessarily the root directory on the drive. That's why I previously suggested that bpo-43105 should be resolved with a `_path_abspath(p)` function that calls `nt._getfullpathname(p)`. It's a relatively cheap call that processes the path as a string. It's not like `nt._getfinalpathname()`, which makes multiple calls into the kernel I/O manager and filesystem.

Reply:
I went with a much simpler implementation that I think still covers all the cases, care to take a look?

Drive relative paths are the ones I'm least concerned about, as they've almost certainly never worked and hence nobody is relying on them. Purely relative paths ought to work though.

Reply:
> Drive relative paths are the ones I'm least concerned about, as they've almost certainly never worked and hence nobody is relying on them.

Drive-relative paths work to some extent prior to 3.8, but only correctly if the path contains a relative path on the drive. Otherwise a bare "X:" will incorrectly use the root directory on the drive. The loader's legacy search with `LOAD_WITH_ALTERED_SEARCH_PATH` allows inserting a drive-relative path in the DLL search path. For example, in 3.7 with "_sqlite3.pyd" and "sqlite3.dll" copied to a test directory:

```python3
>>> sys.path.insert(0, 'C:test37')
>>> import _sqlite3
>>> _sqlite3.__file__
'C:test37\\_sqlite3.pyd'
```

Reply:
Join is just going to leave drive-relative paths as relative now, so I think this will be no-worse than before (while still being better in other cases).